src/Fixer/ReturnNotation/SimplifiedNullReturnFixer.php

```diff
@@ -101,20 +102,51 @@ private function clear(Tokens $tokens, $index)
      */
     private function needFixing(Tokens $tokens, $index)
     {
-        $content = '';
+        if ($this->isStrictOrNullableReturnTypeFunction($tokens, $index)) {
+            return false;
+        }
 
+        $content = '';
         while (!$tokens[$index]->equals(';')) {
             $index = $tokens->getNextMeaningfulToken($index);
             $content .= $tokens[$index]->getContent();
         }
 
-        $content = rtrim($content, ';');
         $content = ltrim($content, '(');
-        $content = rtrim($content, ')');
+        $content = rtrim($content, ');');
 
         return 'null' === strtolower($content);
     }
 
+    /**
+     * Is the return within a function with a non-void or nullable return type?
+     *
+     * @param Tokens $tokens
+     * @param int    $returnIndex Current return token index
+     *
+     * @return bool
+     */
+    private function isStrictOrNullableReturnTypeFunction(Tokens $tokens, $returnIndex)
+    {
+        $functionIndex = $returnIndex;
+        do {
+            $functionIndex = $tokens->getPrevTokenOfKind($functionIndex, array(array(T_FUNCTION)));
+            if (null === $functionIndex) {
+                return false;
+            }
+            $openingCurlyBraceIndex = $tokens->getNextTokenOfKind($functionIndex, array('{'));
+            $closingCurlyBraceIndex = $tokens->findBlockEnd(Tokens::BLOCK_TYPE_CURLY_BRACE, $openingCurlyBraceIndex);
+        } while ($closingCurlyBraceIndex < $returnIndex);
```

Comment:
ha, so that's how you skip lambdas ?

Reply:
finding `closingCurlyBraceIndex` is expensive, we don't care about that information for lambdas, you use it here only to determine if a `T_FUNCTION` a lambda or not.
what about using `TokensAnalyzer::isLambda` instead?

Reply:
The purpose here is different.
Keep in mind that this fixer starts its algorithm from return keyword: what method signature this return belongs to?
Now imagine:

* a class method with a nullable return type
* that contains an anonymous class with a non-nullable method
* that contains an anonymous class with a nullable method
* that contains an anonymous class with a non-nullable method

If in this Matryoshka doll each method also has a return statement, we need this loop to determine each correspondent method signature.

The second-last test case represent this situation.

Reply:
whole method, yes.
But, isn't the `do... while` construction just to skip the content of internal lambda ?

Reply:
No, it's to find out the function's signature the return belongs to.
It goes backward, if the function ends before return skip it (aka keep going backward), if the function ends after return use it.
It doesn't matter which type of function it is.